sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/crypto/_models.py

```diff
@@ -8,6 +8,131 @@
     from . import EncryptionAlgorithm, KeyWrapAlgorithm, SignatureAlgorithm
 
 
+class EncryptionAlgorithmConfiguration():
+    """An encryption algorithm, including parameters applicable to using it for encryption or decryption.
+    """
+
+    def __init__(self, algorithm, **kwargs):
+        self.algorithm = algorithm
+        self.iv = kwargs.pop("iv", None)
+        self.authentication_tag = kwargs.pop("authentication_tag", None)
+        self.additional_authenticated_data = kwargs.pop("additional_authenticated_data", None)
+
+    @classmethod
+    def rsa_oaep_encrypt(cls):
+        return cls(EncryptionAlgorithm.rsa_oaep)
+
+    @classmethod
+    def rsa_oaep_256_encrypt(cls):
+        return cls(EncryptionAlgorithm.rsa_oaep_256)
+
+    @classmethod
+    def rsa1_5_encrypt(cls):
+        return cls(EncryptionAlgorithm.rsa1_5)
+
+    @classmethod
+    def a128_gcm_encrypt(cls, additional_authenticated_data=None):
+        return cls(
+            EncryptionAlgorithm.a128_gcm, additional_authenticated_data=additional_authenticated_data
+        )
+
+    @classmethod
+    def a192_gcm_encrypt(cls, additional_authenticated_data=None):
+        return cls(
+            EncryptionAlgorithm.a192_gcm, additional_authenticated_data=additional_authenticated_data
+        )
+
+    @classmethod
+    def a256_gcm_encrypt(cls, additional_authenticated_data=None):
+        return cls(
+            EncryptionAlgorithm.a256_gcm, additional_authenticated_data=additional_authenticated_data
+        )
+
+    @classmethod
+    def a128_cbc_encrypt(cls, iv=None):
+        return cls(EncryptionAlgorithm.a128_cbc, iv=iv)
+
+    @classmethod
+    def a192_cbc_encrypt(cls, iv=None):
+        return cls(EncryptionAlgorithm.a192_cbc, iv=iv)
+
+    @classmethod
+    def a256_cbc_encrypt(cls, iv=None):
+        return cls(EncryptionAlgorithm.a256_cbc, iv=iv)
+
+    @classmethod
+    def a128_cbcpad_encrypt(cls, iv=None):
+        return cls(EncryptionAlgorithm.a128_cbcpad, iv=iv)
+
+    @classmethod
+    def a192_cbcpad_encrypt(cls, iv=None):
+        return cls(EncryptionAlgorithm.a192_cbcpad, iv=iv)
+
+    @classmethod
+    def a256_cbcpad_encrypt(cls, iv=None):
+        return cls(EncryptionAlgorithm.a256_cbcpad, iv=iv)
+
+    @classmethod
+    def rsa_oaep_decrypt(cls):
+        return cls(EncryptionAlgorithm.rsa_oaep)
+
+    @classmethod
+    def rsa_oaep_256_decrypt(cls):
+        return cls(EncryptionAlgorithm.rsa_oaep_256)
+
+    @classmethod
+    def rsa1_5_decrypt(cls):
+        return cls(EncryptionAlgorithm.rsa1_5)
+
+    @classmethod
+    def a128_gcm_decrypt(cls, authentication_tag=None, additional_authenticated_data=None):
```

Comment:
This gives me pause. The algorithm is the same regardless of whether you're encrypting or decrypting but the tag is an encryption product only used in decryption. Having two methods for the same algorithm is weird (doubly so when their signatures are identical) but combining them entails asking someone about to encrypt for an optional useless argument, which is contrary to our purpose. Having `_encrypt` and `_decrypt` methods for only some algorithms is also weird, as is adding a `DecryptionAlgorithmConfiguration` class. Well, that last thought isn't totally out there. But all this has me thinking we need some more ideas here 🤔

Reply:
The last option would be most similar to what other languages are doing with `EncryptParameters` and `DecryptParameters`. I can't recall why I was opposed to having `EncryptionAlgorithmConfiguration` and `DecryptionAlgorithmConfiguration` -- maybe because they both take `EncryptionAlgorithm`s, and the latter obfuscates that -- but it's definitely one of the better options